File: agent_hub/orchestrator.py
import os
import json
import anthropic
from agents import ClaudeAgent, OpenAIAgent, GeminiAgent

ALL_AGENTS = [ClaudeAgent, OpenAIAgent, GeminiAgent]

ORCHESTRATOR_SYSTEM = """You are an AI orchestrator. Your job is to:
1. Analyze the user's task
2. Decide which AI agent(s) would produce the best result
3. Optionally split the task into sub-tasks for different agents
4. Synthesize their outputs into a single finished product

Available agents and their strengths:
- Claude: strong reasoning, code, structured analysis, technical writing
- ChatGPT: creative writing, broad general knowledge, conversational tone
- Gemini: research synthesis, long-form content, factual lookup

Respond ONLY with valid JSON in this exact format:
{
  "plan": "one sentence explaining your approach",
  "assignments": [
    {
      "agent": "<Claude|ChatGPT|Gemini>",
      "subtask": "the specific task for this agent",
      "context": "any extra context or instructions for this agent"
    }
  ],
  "synthesis_instruction": "how to combine results into the final output (or 'passthrough' if only one agent)"
}

Only assign agents that are listed as available. Always produce a complete, finished result."""


def plan_and_execute(task: str, available_agents: list[str]) -> dict:
    """
    Orchestrate: plan via Claude, execute subtasks, synthesize results.
    Returns a dict with plan, agent_results, and final_output.
    """
    client = anthropic.Anthropic(api_key=os.getenv("ANTHROPIC_API_KEY"))

    available_str = ", ".join(available_agents)
    planning_prompt = f"Available agents: {available_str}\n\nUser task:\n{task}"

    plan_response = client.messages.create(
        model="claude-sonnet-4-6",
        max_tokens=1024,
        system=ORCHESTRATOR_SYSTEM,
        messages=[{"role": "user", "content": planning_prompt}],
    )

    raw = plan_response.content[0].text.strip()
    # Strip markdown code fences if present
    if raw.startswith("```"):
        raw = raw.split("```")[1]
        if raw.startswith("json"):
            raw = raw[4:]
    plan = json.loads(raw)

    agent_map = {
        "Claude": ClaudeAgent,
        "ChatGPT": OpenAIAgent,
        "Gemini": GeminiAgent,
    }

    agent_results = []
    for assignment in plan["assignments"]:
        agent_name = assignment["agent"]
        if agent_name not in available_agents:
            continue
        agent_cls = agent_map.get(agent_name)
        if not agent_cls:
            continue

        result = agent_cls.run(assignment["subtask"], assignment.get("context", ""))
        agent_results.append({
            "agent": agent_name,
            "subtask": assignment["subtask"],
            "result": result,
        })

    if len(agent_results) == 0:
        return {
            "plan": plan.get("plan", ""),
            "agent_results": [],
            "final_output": "No agents were available to complete this task.",
        }

    # Single agent — return directly
    if len(agent_results) == 1 or plan.get("synthesis_instruction") == "passthrough":
        return {
            "plan": plan.get("plan", ""),
            "agent_results": agent_results,
            "final_output": agent_results[0]["result"],
        }

    # Multiple agents — synthesize
    synthesis_parts = []
    for r in agent_results:
        synthesis_parts.append(f"=== {r['agent']} output ===\n{r['result']}")

    synthesis_prompt = (
        f"Original task: {task}\n\n"
        f"Synthesis instruction: {plan['synthesis_instruction']}\n\n"
        + "\n\n".join(synthesis_parts)
        + "\n\nCombine the above into one cohesive, finished result for the user."
    )

    synthesis_response = client.messages.create(
        model="claude-sonnet-4-6",
        max_tokens=4096,
        system="You are a synthesis AI. Combine multiple AI outputs into a single polished, finished result.",
        messages=[{"role": "user", "content": synthesis_prompt}],
    )

    return {
        "plan": plan.get("plan", ""),
        "agent_results": agent_results,
        "final_output": synthesis_response.content[0].text,
    }

File: agent_hub_app/src/orchestrator.js
const { runClaude } = require("./agents/claude");
const { runOpenAI } = require("./agents/openai");
const { runGemini } = require("./agents/gemini");

const ORCHESTRATOR_SYSTEM = `You are an AI orchestrator. Analyze the user's task and decide which AI agent(s) should handle it.

Agent strengths:
- Claude: reasoning, code, structured analysis, technical writing
- ChatGPT: creative writing, general knowledge, conversational tasks
- Gemini: research, long-form content, factual synthesis

Respond ONLY with valid JSON (no markdown fences):
{
  "plan": "one sentence explaining your approach",
  "assignments": [
    {
      "agent": "Claude|ChatGPT|Gemini",
      "subtask": "specific task for this agent",
      "context": "extra context or instructions"
    }
  ],
  "synthesis_instruction": "how to combine results, or 'passthrough' if only one agent"
}

Only assign agents listed as available. Always produce a complete, finished result.`;

async function orchestrate(task, { anthropicKey, openaiKey, googleKey }) {
  const available = [];
  if (anthropicKey) available.push("Claude");
  if (openaiKey) available.push("ChatGPT");
  if (googleKey) available.push("Gemini");

  // Plan
  const Anthropic = require("@anthropic-ai/sdk");
  const client = new Anthropic.default({ apiKey: anthropicKey });

  const planMsg = await client.messages.create({
    model: "claude-sonnet-4-6",
    max_tokens: 1024,
    system: ORCHESTRATOR_SYSTEM,
    messages: [
      {
        role: "user",
        content: `Available agents: ${available.join(", ")}\n\nUser task:\n${task}`,
      },
    ],
  });

  let raw = planMsg.content[0].text.trim();
  if (raw.startsWith("```")) {
    raw = raw.replace(/^```[a-z]*\n?/, "").replace(/\n?```$/, "");
  }
  const plan = JSON.parse(raw);

  // Execute
  const agentResults = [];
  for (const assignment of plan.assignments) {
    if (!available.includes(assignment.agent)) continue;
    let result;

    if (assignment.agent === "Claude") {
      result = await runClaude(assignment.subtask, assignment.context || "", anthropicKey);
    } else if (assignment.agent === "ChatGPT") {
      result = await runOpenAI(assignment.subtask, assignment.context || "", openaiKey);
    } else if (assignment.agent === "Gemini") {
      result = await runGemini(assignment.subtask, assignment.context || "", googleKey);
    }

    agentResults.push({ agent: assignment.agent, subtask: assignment.subtask, result });
  }

  if (agentResults.length === 0) {
    return { plan: plan.plan, agentResults: [], finalOutput: "No agents available." };
  }

  // Single agent or passthrough
  if (agentResults.length === 1 || plan.synthesis_instruction === "passthrough") {
    return { plan: plan.plan, agentResults, finalOutput: agentResults[0].result };
  }

  // Synthesize
  const parts = agentResults.map((r) => `=== ${r.agent} ===\n${r.result}`).join("\n\n");
  const synthesisPrompt =
    `Original task: ${task}\n\nInstruction: ${plan.synthesis_instruction}\n\n${parts}\n\n` +
    "Combine the above into one cohesive, finished result.";

  const synthMsg = await client.messages.create({
    model: "claude-sonnet-4-6",
    max_tokens: 4096,
    system: "You are a synthesis AI. Combine multiple AI outputs into one polished, finished result.",
    messages: [{ role: "user", content: synthesisPrompt }],
  });

  return {
    plan: plan.plan,
    agentResults,
    finalOutput: synthMsg.content[0].text,
  };
}

module.exports = { orchestrate };
